### PR TITLE
Add PgBouncer Client

### PR DESCRIPTION
**Pull request by rakibulhossain**
Signed-off-by: Rakibul-Hossain <rakibul.hossain@appscode.com>

### CODE REVIEW

File: pgbouncer/kubedb_client_builder.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+func (o *KubeDBClientBuilder) getURL() string {
+	return fmt.Sprintf("%s.%s.%s.svc", o.podName, o.db.GoverningServiceName(), o.db.Namespace)
+}
+
+func (o *KubeDBClientBuilder) getPgBouncerAuthCredentials() (string, string, error) {
+	if o.db.Spec.AuthSecret == nil {
+		return "", "", errors.New("no database secret")
+	}
+	var secret core.Secret
+	err := o.kc.Get(context.TODO(), client.ObjectKey{Namespace: o.db.Namespace, Name: o.db.Spec.AuthSecret.Name}, &secret)
```
Comment (pkbhowmick) [inline pgbouncer/kubedb_client_builder.go:98]: better to take context in method param and use it here

File: pgbouncer/kubedb_client_builder.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		return nil, err
+	}
+
+	// ping to database to check the connection
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	if _, err := db.QueryContext(ctx, "SHOW HELP;"); err != nil {
+		klog.Infoln(db)
+		klog.Infoln("here is error", connector)
```
Comment (pkbhowmick) [inline pgbouncer/kubedb_client_builder.go:122]: can we improve error msg? better to use klog.Errorf() or klog.Error()

File: pgbouncer/kubedb_client_builder.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		log.Fatal(err)
```
Comment (pkbhowmick) [inline pgbouncer/kubedb_client_builder.go:123]: Avoid log.Fatal(). We can return error from here.

File: pgbouncer/kubedb_client_builder.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+	return &Client{db}, nil
+}
+
+func (o *KubeDBClientBuilder) getConnectionString() (string, error) {
+	if o.podName != "" {
+		o.url = o.getURL()
+	}
+	dnsName := o.url
+	port := 5432
```
Comment (pkbhowmick) [inline pgbouncer/kubedb_client_builder.go:134]: Use constant for port

File: pgbouncer/kubedb_client_builder.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		return "", fmt.Errorf("DB basic auth is not found for PgBouncer %v/%v", o.db.Namespace, o.db.Name)
+	}
+	cnnstr := ""
+	sslMode := o.db.Spec.SSLMode
+	if sslMode == "" {
+		sslMode = api.PgBouncerSSLModeDisable
+	}
+
+	if o.db.Spec.TLS != nil {
+		secretName := o.db.GetCertSecretName(api.PgBouncerClientCert)
```
Comment (pkbhowmick) [inline pgbouncer/kubedb_client_builder.go:151]: can we refactor this part into a different method ? Like `func (o *KubeDBClientBuilder) getTLSConfig() (tlsConfig, error)`. Then use it in building connection string.

File: pgbouncer/kubedb_client_builder.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+		var certSecret core.Secret
+		err := o.kc.Get(context.TODO(), client.ObjectKey{Namespace: o.db.Namespace, Name: secretName}, &certSecret)
```
Comment (pkbhowmick) [inline pgbouncer/kubedb_client_builder.go:154]: use context that will come from param.

**Review by pkbhowmick — CHANGES_REQUESTED** [review]

File: pgbouncer/kubedb_client_builder.go (hunk 2 above)
Comment (rakibulhossain) [reply pgbouncer/kubedb_client_builder.go:122]: It's for dbg, forgot to remove

**Review by pkbhowmick — DISMISSED** [review]
LGTM!

**Review by pkbhowmick — APPROVED** [review]

**Pull request merged**